### PR TITLE
Fixing sticky immix

### PR DESCRIPTION
**Pull request by udesou**
This PR should be merged with https://github.com/mmtk/julia/pull/32. 

Copying the description here for simplicity:

> At least one of the issues with sticky immix back ported to [v1.9.2+RAI](https://github.com/mmtk/mmtk-julia/tree/v1.9.2%2BRAI) is that bindings [may occur in the remset](https://github.com/mmtk/julia/blob/c01026c91a3e2e6a064e75e0d4d4cc2f8c0d4c77/src/gc.c#L2417-L2433). However, bindings are currently skipped when scanned by MMTk since they are allocated as buffers. Therefore, we need a way to specify which bindings should actually be scanned since they were added via write barrier.
> 
> Possible sources of problems that haven't been verified yet:
> (i) can the `bits.gc` value change after being set by the write barrier?
> (ii) should we reset the value after the object has been scanned?

This PR adds code for scanning buffers that are Julia bindings that were added in the write barrier.

### CODE REVIEW

File: mmtk/src/julia_scanning.rs [hunk 1]
```diff
@@ -43,7 +43,28 @@ pub unsafe fn scan_julia_object<EV: EdgeVisitor<JuliaVMEdge>>(obj: Address, clos
     // get Julia object type
     let vt = mmtk_jl_typeof(obj);
 
-    if vt == jl_symbol_type || vt as usize == JULIA_BUFF_TAG {
+    if vt as usize == JULIA_BUFF_TAG {
+        let as_tagged_value =
+            obj.as_usize() - std::mem::size_of::<crate::julia_scanning::mmtk_jl_taggedvalue_t>();
+        let t_header = Address::from_usize(as_tagged_value).load::<Address>();
+        let tag = t_header.as_usize() & 3;
+        if tag == 2 {
```
Comment (qinsoon) [inline mmtk/src/julia_scanning.rs:51]: I can see from the Julia code, that the write barrier for `jl_binding_t` sets the bits to 2. It means if `jl_binding_t` gets modified, the write barrier will set the bits to 2, and we will trace the fields in the `jl_binding_t`. You probably explained to me earlier why we need to do this, and why we don't need to trace fields in `jl_binding_t` if it is not in the remset. But I forgot the actual reason.

Can you write it into the comments why we need this code?

Comment (qinsoon) [reply mmtk/src/julia_scanning.rs:51]: Is the bits cleared in any way? If not, that means once we sets the bits for a `jl_binding_t` object, the bits stay as 2, and we will trace its fields in every subsequent GCs.

Comment (qinsoon) [reply mmtk/src/julia_scanning.rs:51]: > I can see from the Julia code, that the write barrier for `jl_binding_t` sets the bits to 2. It means if `jl_binding_t` gets modified, the write barrier will set the bits to 2, and we will trace the fields in the `jl_binding_t`. You probably explained to me earlier why we need to do this, and why we don't need to trace fields in `jl_binding_t` if it is not in the remset. But I forgot the actual reason.
> 
> Can you write it into the comments why we need this code?

Oh I see. We don't scan buffers, as they will be scanned as a part of its parent object. But when a `jl_binding_t` buffer is inserted into remset, they have be to scanned. The gc bits tells us if the buffer is in the remset. Can you put this into comments?

Comment (udesou) [reply mmtk/src/julia_scanning.rs:55]: > Is the bits cleared in any way? If not, that means once we sets the bits for a `jl_binding_t` object, the bits stay as 2, and we will trace its fields in every subsequent GCs.

I've added a `FIXME` for now, but that's true. We should reset the bits after the `jl_binding_t` object has been scanned.

Comment (qinsoon) [reply mmtk/src/julia_scanning.rs:55]: Why not just clear the bits after the `process_edge` calls? Are you worrying that they will cause other issues? If they do (for example, if those bits are used by others for other purposes), it means the implementation is not sound.

**Review by qinsoon — APPROVED** [review]

**Comment by qinsoon** [comment]
This PR only affects v1.9.2, not `master`.

> Eduardo: they don't use buffers to store bindings in master. And IIRC they don't have the special write barrier either
> Eduardo: I think they store them in a plain `jl_array_t`

**Pull request merged**